### PR TITLE
Fix float32 overflow in variance computation for normalization layers

### PR DESCRIPTION
**Pull request by gaurav-redhat**
Fixes #162489

## Problem

When float32 inputs have large magnitude (~1e30), the variance computation in BatchNorm, LayerNorm, GroupNorm, and InstanceNorm overflows. The Welford online algorithm squares the deviation `(x - mean)`, and `1e30 * 1e30 = 1e60` which exceeds float32 max (~3.4e38). This causes:

- **CPU**: variance becomes `inf`, normalization produces all zeros
- **CUDA**: variance becomes `inf`, `inf - inf = NaN`, normalization produces NaN

## Fix

Promote the Welford accumulation variables (`mean`, `m2`) from `float` to `double` when the input is float32. The rest of the computation stays in float32 — only the running sums that can overflow are widened.

This is done using `std::conditional_t<std::is_same_v<T_ACC, float>, double, T_ACC>` so it's zero-cost for float64 and lower-precision types.

### CPU kernels
- `batch_norm_kernel.cpp` — all 4 variance paths (contiguous, channels-last method 1/2, large-N fallback)
- `group_norm_kernel.cpp` — `ColumnwiseMoments` and `CalcMeanVar`
- `moments_utils.h` — scalar Welford fallback in `RowwiseMomentsImpl` (used by LayerNorm and GroupNorm)

### CUDA kernels
- `Normalization.cuh` — BatchNorm contiguous and channels-last stats kernels
- `layer_norm_kernel.cu` — `RowwiseMomentsCUDAKernel` and vectorized `WelfordDataLN` path
- `group_norm_kernel.cu` — `RowwiseMomentsCUDAKernel`

## Test

Added `test_norm_extreme_values_no_overflow` in `test_nn.py` that tests all norm layers with `torch.linspace(-1e30, 1e30, ...)` on both CPU and CUDA — verifies no NaN/Inf/zero in the output.

## Repro (from the issue)

```python
import torch
from torch.nn import BatchNorm1d

x = torch.linspace(-1e30, 1e30, steps=4).reshape(2, 2).to('cuda')
out = BatchNorm1d(2, affine=False).to('cuda')(x)
print(out)  # Before: tensor([[nan, nan], [nan, nan]])
            # After:  tensor([[-1., -1.], [ 1.,  1.]])
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
Is this a real use-case? It seems like it would have catastrophic performance impact
